monai/networks/nets/dynunet.py

```diff
@@ -180,30 +180,52 @@ def create_skips(index, downsamples, upsamples, superheads, bottleneck):
 
             if len(downsamples) != len(upsamples):
                 raise ValueError(f"{len(downsamples)} != {len(upsamples)}")
-            if (len(downsamples) - len(superheads)) not in (1, 0):
-                raise ValueError(f"{len(downsamples)}-(0,1) != {len(superheads)}")
 
             if len(downsamples) == 0:  # bottom of the network, pass the bottleneck block
                 return bottleneck
-            if index == 0:  # don't associate a supervision head with self.input_block
-                current_head, rest_heads = nn.Identity(), superheads
-            elif not self.deep_supervision:  # bypass supervision heads by passing nn.Identity in place of a real one
-                current_head, rest_heads = nn.Identity(), superheads[1:]
-            else:
-                current_head, rest_heads = superheads[0], superheads[1:]
-
-            # create the next layer down, this will stop at the bottleneck layer
-            next_layer = create_skips(1 + index, downsamples[1:], upsamples[1:], rest_heads, bottleneck)
-
-            return DynUNetSkipLayer(index, self.heads, downsamples[0], upsamples[0], current_head, next_layer)
 
-        self.skip_layers = create_skips(
-            0,
-            [self.input_block] + list(self.downsamples),
-            self.upsamples[::-1],
-            self.deep_supervision_heads,
-            self.bottleneck,
-        )
+            if superheads is None:
+                next_layer = create_skips(1 + index, downsamples[1:], upsamples[1:], bottleneck)
+                return DynUNetSkipLayer(index, downsample=downsamples[0], upsample=upsamples[0], next_layer=next_layer)
+            else:
+                super_head_flag = False
+                if index == 0:  # don't associate a supervision head with self.input_block
+                    rest_heads = superheads
+                else:
+                    if len(superheads) > 0:
+                        super_head_flag = True
+                        rest_heads = superheads[1:]
+                    else:
+                        rest_heads = nn.ModuleList()
+
+                # create the next layer down, this will stop at the bottleneck layer
+                next_layer = create_skips(1 + index, downsamples[1:], upsamples[1:], bottleneck, rest_heads)
+                if super_head_flag:
+                    return DynUNetSkipLayer(
+                        index,
+                        downsample=downsamples[0],
+                        upsample=upsamples[0],
+                        next_layer=next_layer,
+                        heads=self.heads,
+                        super_head=superheads[0],
+                    )
+                else:
+                    return DynUNetSkipLayer(
+                        index, downsample=downsamples[0], upsample=upsamples[0], next_layer=next_layer
+                    )
+
+        if not self.deep_supervision:
```

Comment:
Is this `if-else` block reachable? The above `if-else` block has returns in both parts, the latter in a `if-else` based on the boolean flag `super_head_flag`. I think in any possibility this will hit a return unless I missed something.

Reply:
Hi @ericspod , thanks for the review, the above `if-else` block is actually in `def create_skips`